File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/designsystem/AtomsDoNotComposeOtherAtoms.kt
package com.perrystreet.woof.konsist.designsystem

import com.lemonappdev.konsist.api.ext.list.withAnnotationNamed
import com.perrystreet.woof.konsist.Assertions.assertFalse
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.KonsistUtils.body
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class AtomsDoNotComposeOtherAtoms : BehaviorSpec() {
    init {
        Given("An atom composable") {
            val atoms = KonsistUtils.atomicDesignModule.functions().withAnnotationNamed("Composable").filter { it.path.contains("/atomic/atoms/") }

            Then("It does not call another atom") {
                atoms.assertFalse(message = Message) { atom -> AtomCallRegex.containsMatchIn(atom.body) }
            }
        }
    }

    private companion object {
        private val AtomCallRegex = Regex("""\bAtom[A-Z]\w*\(""")

        private val Message = LintRuleMessage(
            rule = "Atoms are built from Compose primitives only; they never call other atoms.",
            why = "The moment two atoms combine, the result is a molecule. Keeping atoms leaf-level keeps the hierarchy honest and the dependency graph flat.",
            howToFix = "Promote the composition to a Mol* component in atomic/molecules.",
            badExample = "@Composable fun AtomLabeledIcon() { AtomIcon(...); AtomText(...) }",
            goodExample = "@Composable fun MolLabeledIcon() { AtomIcon(...); AtomText(...) }",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/DesignSystem/AtomsDoNotComposeOtherAtoms.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class AtomsDoNotComposeOtherAtoms: QuickSpec {
    override class func spec() {
        Given("An atom view") {
            let atoms = WoofHarmonize.atomicDesignPackage.structs().views.inFolder("Atoms")

            Then("It does not call another atom") {
                atoms.assertFalse(rule: rule) { atom in
                    atom.description.containsMatch(of: atomCallPattern)
                }
            }
        }
    }

    private static let atomCallPattern = try! NSRegularExpression(pattern: #"\bAtom[A-Z]\w*\("#)

    private static let rule = Rule(
        description: "Atoms are built from SwiftUI primitives only; they never call other atoms.",
        rationale: "The moment two atoms combine, the result is a molecule. Keeping atoms leaf-level keeps the hierarchy honest and the dependency graph flat.",
        fixHint: "Promote the composition to a Mol* component in Atomic/Molecules.",
        badExample: "struct AtomLabeledIcon: View { var body: some View { HStack { AtomIcon(...); AtomText(...) } } }",
        goodExample: "struct MolLabeledIcon: View { var body: some View { HStack { AtomIcon(...); AtomText(...) } } }"
    )
}
